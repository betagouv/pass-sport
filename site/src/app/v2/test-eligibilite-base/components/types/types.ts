export enum AGE_RANGE {
  BETWEEN_6_19 = 'BETWEEN_6_19',
  BETWEEN_19_30 = 'BETWEEN_19_30',
  GREATER_THAN_30 = 'GREATER_THAN_30',
}

export enum CHILD_AGE {
  LESS_THAN_SIX = 'LESS_THAN_SIX',
  BTW_SIX_AND_THIRTY = 'BTW_SIX_AND_THIRTY',
  MORE_THAN_THIRTY = 'MORE_THAN_THIRTY',
}
