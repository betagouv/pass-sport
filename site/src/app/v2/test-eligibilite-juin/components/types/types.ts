export enum ALLOWANCE {
  NONE = 'NONE',
  ARS_AEEH_AAH = 'ARS_AEEH_AAH',
  CROUS = 'CROUS',
}
