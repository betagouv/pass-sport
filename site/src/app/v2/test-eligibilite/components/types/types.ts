export enum AGE_RANGE {
  BETWEEN_6_19 = 'BETWEEN_6_19',
  BETWEEN_19_30 = 'BETWEEN_19_30',
  GREATER_THAN_30 = 'GREATER_THAN_30',
}

export enum CHILD_AGE {
  LESS_THAN_SIX = 'lessThanSix',
  BTW_SIX_AND_THIRTY = 'betweenSixAndThrity',
  MORE_THAN_THIRTY = 'moreThanThirty',
}
