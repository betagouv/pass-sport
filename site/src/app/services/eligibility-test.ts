import {
  ConfirmPayload,
  ConfirmResponseBody,
  ConfirmResponseErrorBody,
  SearchPayload,
  SearchResponseBody,
  SearchResponseErrorBody,
} from 'types/EligibilityTest';

import * as Sentry from '@sentry/nextjs';
import { addQrCodeToConfirmResponse } from './qr-code';

export const buildLCAConfirmUrl = (data: ConfirmPayload, isUsingApiV1: boolean): URL => {
  const domain = process.env.LCA_API_URL;

  if (!domain) {
    throw new Error('Error: LCA_API_URL is not set');
  }

  const params = new URLSearchParams();
  params.append('id', data.id);
  params.append('situation', data.situation);
  params.append('organisme', data.organisme);

  const allocataireName = data.recipientLastname;
  if (allocataireName) {
    params.append('allocataireName', allocataireName);
  }

  const allocataireSurname = data.recipientFirstname;
  if (allocataireSurname) {
    params.append('allocataireSurname', allocataireSurname);
  }

  const matricule = data.recipientCafNumber;
  if (matricule) {
    params.append('matricule', matricule);
  }

  const recipientBirthPlace = data.recipientBirthPlace;
  if (recipientBirthPlace) {
    params.append('codeInseeBirth', recipientBirthPlace);
  }

  const recipientBirthDate = data.recipientBirthDate;
  if (recipientBirthDate) {
    params.append('allocataireBirthDate', recipientBirthDate);
  }

  const recipientBirthCountry = data.recipientBirthCountry;
  if (recipientBirthCountry) {
    params.append('codeIso', recipientBirthCountry);
  }

  const baseUrl = isUsingApiV1
    ? `${domain}/gw/psp-server/beneficiaires/confirm`
    : `${domain}/apim/api-asso-admin/passsport/beneficiaires/confirm`;

  const url = new URL(baseUrl);
  url.search = params.toString();

  return url;
};

export const buildLCASearchUrl = (data: SearchPayload, isUsingApiV1: boolean): URL => {
  const domain = process.env.LCA_API_URL;

  if (!domain) {
    throw new Error('Error: LCA_API_URL is not set');
  }

  const params = new URLSearchParams();
  params.append('nom', data.beneficiaryLastname);
  params.append('prenom', data.beneficiaryFirstname);
  params.append('dateNaissance', data.beneficiaryBirthDate);
  params.append('codeInsee', data.recipientResidencePlace);

  const baseUrl = isUsingApiV1
    ? `${domain}/gw/psp-server/beneficiaires/search`
    : `${domain}/apim/api-asso-admin/passsport/beneficiaires/search`;

  const url = new URL(baseUrl);
  url.search = params.toString();

  return url;
};

export const fetchQrCode = async (
  payload: ConfirmPayload,
): Promise<ConfirmResponseBody | ConfirmResponseErrorBody> => {
  const authenticationKey = process.env.LCA_API_KEY;

  if (!authenticationKey) {
    throw new Error('LCA authentication key is missing');
  }

  const url: URL = buildLCAConfirmUrl(payload, true);
  const url2 = buildLCAConfirmUrl(payload, false);

  const [response1, response2] = await Promise.all([
    fetch(url),
    fetch(url2, { headers: { 'X-Gravitee-Api-Key': authenticationKey } }),
  ]);

  if (!response2.ok && !response1.ok) {
    throw new Error(
      `Request to LCA api on /confirm has failed. API V1: Response status is ${response1.status}; Response body is ${JSON.stringify(await response1.json())}. API V2: Response status is ${response2.status}; Response body is ${JSON.stringify(await response2.json())}`,
    );
  }

  const response = !response2.ok ? response1 : response2;
  const responseBody = (await response.json()) as ConfirmResponseBody | ConfirmResponseErrorBody;

  if ('message' in responseBody) {
    Sentry.withScope((scope) => {
      scope.setLevel('warning');
      scope.setExtra('responseBody', responseBody);
      scope.captureMessage('Unexpected response on LCA POST api/eligibility-test/confirm');
    });
    return responseBody;
  }
  if (responseBody instanceof Array && responseBody.length === 0) {
    return responseBody;
  }
  const enhancedResponse = addQrCodeToConfirmResponse(responseBody);
  return enhancedResponse;
};

export const fetchEligible = async (payload: SearchPayload) => {
  const authenticationKey = process.env.LCA_API_KEY;

  if (!authenticationKey) {
    throw new Error('LCA authentication key is missing');
  }

  const url: URL = buildLCASearchUrl(payload, true);
  const url2: URL = buildLCASearchUrl(payload, false);

  const [response1, response2] = await Promise.all([
    fetch(url),
    fetch(url2, { headers: { 'X-Gravitee-Api-Key': authenticationKey } }),
  ]);

  if (!response2.ok && !response1.ok) {
    throw new Error(
      `Request to LCA api on /search has failed. API V1: Response status is ${response1.status}; Response body is ${JSON.stringify(await response1.json())}. API V2: Response status is ${response2.status}; Response body is ${JSON.stringify(await response2.json())}`,
    );
  }

  const response = !response2.ok ? response1 : response2;
  const responseBody = (await response.json()) as SearchResponseBody | SearchResponseErrorBody;

  if ('message' in responseBody) {
    Sentry.withScope((scope) => {
      scope.setLevel('warning');
      scope.setExtra('responseBody', responseBody);
      scope.captureMessage('Unexpected response on LCA POST api/eligibility-test/search');
    });
  }

  return responseBody;
};
