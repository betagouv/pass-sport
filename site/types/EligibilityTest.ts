type InputState = 'default' | 'success' | 'error';

export interface InputsState {
  beneficiaryLastname: InputState;
  beneficiaryFirstname: InputState;
  beneficiaryBirthDate: InputState;
  recipientResidencePlace: InputState;
}

export interface SearchResponseBody {
  id: number;
  nom: string;
  prenom: string;
  date_naissance: string;
  situation: string;
  organisme: string;
  matricule: string;
}

export interface SearchResponseError {
  message: string;
}
