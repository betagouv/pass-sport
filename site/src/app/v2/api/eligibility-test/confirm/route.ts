import { buildLCAConfirmUrl } from '@/app/services/eligibility-test';
import { addQrCodeToConfirmResponse } from '@/app/services/qr-code';
import { NextResponse } from 'next/server';
import {
  ConfirmPayload,
  ConfirmResponseBody,
  ConfirmResponseErrorBody,
} from 'types/EligibilityTest';
import { zfd } from 'zod-form-data';
import z from 'zod';

const schema = zfd.formData({
  id: z.string(),
  situation: z.string(),
  organisme: z.string(),
  recipientLastname: z.string().optional(),
  recipientFirstname: z.string().optional(),
  recipientCafNumber: z.string().optional(),
  recipientBirthPlace: z.string().optional(),
  recipientBirthDate: z.string().optional(),
  recipientBirthCountry: z.string().optional(),
});

export async function POST(request: Request) {
  try {
    const formData = await request.formData();
    const payload: ConfirmPayload = schema.parse(formData);

    const url: URL = buildLCAConfirmUrl(payload);
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(
        `Request to LCA api on /confirm has failed. Response status is ${response.status}`,
      );
    }

    const responseBody = (await response.json()) as ConfirmResponseBody | ConfirmResponseErrorBody;

    if ('message' in responseBody) {
      console.error(
        'POST api/eligibility-test/confirm',
        'request to LCA has a message field in the body. The body is: ',
        responseBody,
      );
      return NextResponse.json(responseBody);
    }

    const enhancedResponse = addQrCodeToConfirmResponse(responseBody);
    return NextResponse.json(enhancedResponse);
  } catch (e) {
    console.error('POST api/eligibility-test/confirm', 'internal error', e);
    return NextResponse.json('Internal error', { status: 500 });
  }
}
