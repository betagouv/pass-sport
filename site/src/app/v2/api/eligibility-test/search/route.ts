import { fetchEligible } from '@/app/services/eligibility-test';
import { SearchPayload } from 'types/EligibilityTest';
import { zfd } from 'zod-form-data';
import z, { ZodError } from 'zod';
import * as Sentry from '@sentry/nextjs';
import { handleSupportCookie } from '@/utils/cookie';

const schema = zfd.formData({
  beneficiaryLastname: z.string(),
  beneficiaryFirstname: z.string(),
  beneficiaryBirthDate: z.string(),
  recipientResidencePlace: z.string(),
});

export async function POST(request: Request): Promise<Response> {
  try {
    const formData = await request.formData();
    const payload: SearchPayload = schema.parse(formData);

    const data = await fetchEligible(payload);

    // Means no one was found
    if (Array.isArray(data) && data.length <= 0) {
      await handleSupportCookie(payload, 'search');
    }

    return Response.json(data);
  } catch (e) {
    if (e instanceof ZodError) {
      return Response.json('Some fields are missing', { status: 400 });
    }

    Sentry.withScope((scope) => {
      scope.setLevel('error');
      scope.captureMessage('Technical error on LCA POST api/eligibility-test/search');
      scope.captureException(e);
    });

    return Response.json('Internal error', { status: 500 });
  }
}
