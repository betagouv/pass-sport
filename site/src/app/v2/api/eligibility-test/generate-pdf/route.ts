import { zfd } from 'zod-form-data';
import z, { ZodError } from 'zod';
import { generatePdfBuffer } from '@/app/v2/api/eligibility-test/confirm/generate-pdf-buffer';
import { NextResponse } from 'next/server';

const schema = zfd.formData({
  lastname: z.string(),
  firstname: z.string(),
  dob: z.string(),
  code: z.string(),
  gender: z.enum(['M', 'F']),
});

export async function POST(request: Request): Promise<Response> {
  try {
    const { lastname, firstname, dob, gender, code } = schema.parse(await request.formData());

    const pdfBuffer = await generatePdfBuffer({
      lastname,
      firstname,
      dob,
      gender,
      code,
    });

    // 3. Set the response headers
    const headers = new Headers();

    headers.set('Content-Type', 'application/pdf');
    headers.set(
      'Content-Disposition',
      `attachment; filename="code-pass-sport${lastname}-${firstname}.pdf"`,
    );
    headers.set('Content-Length', pdfBuffer.length.toString());

    return new NextResponse(pdfBuffer, { status: 200, headers });
  } catch (e) {
    if (e instanceof ZodError) {
      return Response.json('Some fields are missing', { status: 400 });
    }

    return Response.json('Internal error', { status: 500 });
  }
}
