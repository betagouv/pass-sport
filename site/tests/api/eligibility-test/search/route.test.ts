/**
 * @jest-environment node
 */

import { fetchEligible } from '@/app/services/eligibility-test';
import { POST } from '@/app/v2/api/eligibility-test/search/route';
import { buildSearchResponseBody } from '../../../../tests/helpers/builders/confirm-response-body';
import { SearchResponseBody } from 'types/EligibilityTest';

jest.mock('../../../../src/app/services/eligibility-test', () => {
  const original = jest.requireActual('../../../../src/app/services/eligibility-test');
  return {
    ...original,
    fetchEligible: jest.fn(),
  };
});

describe('POST /eligibility-test/search', () => {
  const badRequestTests = [
    {
      msg: 'beneficiaryLastname is not provided',
      data: [
        { key: 'beneficiaryFirstname', value: 'bob' },
        { key: 'beneficiaryBirthDate', value: '2015-03-28' },
        { key: 'recipientResidencePlace', value: '05024' },
      ],
    },
    {
      msg: 'beneficiaryFirstname is not provided',
      data: [
        { key: 'beneficiaryLastname', value: 'marley' },
        { key: 'beneficiaryBirthDate', value: '2015-03-28' },
        { key: 'recipientResidencePlace', value: '05024' },
      ],
    },
    {
      msg: 'beneficiaryBirthDate is not provided',
      data: [
        { key: 'beneficiaryFirstname', value: 'bob' },
        { key: 'beneficiaryLastname', value: 'marley' },
        { key: 'recipientResidencePlace', value: '05024' },
      ],
    },
    {
      msg: 'recipientResidencePlace is not provided',
      data: [
        { key: 'beneficiaryFirstname', value: 'bob' },
        { key: 'beneficiaryLastname', value: 'marley' },
        { key: 'beneficiaryBirthDate', value: '2015-03-28' },
      ],
    },
  ];

  it.each(badRequestTests)('return a 400 when $msg ', async ({ data }) => {
    const payload = new FormData();
    data.forEach((item) => {
      payload.append(item.key, item.value);
    });

    const request = new Request('http://localhost/api/eligibility-test/search', {
      method: 'POST',
      body: payload,
    });

    const response = await POST(request);
    expect(response.status).toEqual(400);
  });

  it('should return fetched data', async () => {
    const payload = new FormData();
    payload.append('beneficiaryLastname', 'DUPOND');
    payload.append('beneficiaryFirstname', 'MANON');
    payload.append('beneficiaryBirthDate', '2011-01-01');
    payload.append('recipientResidencePlace', '05024');

    const request = new Request('http://localhost/api/eligibility-test/search', {
      method: 'POST',
      body: payload,
    });

    const mockedFetchEligible = fetchEligible as jest.Mock;
    const mockedResponse: SearchResponseBody = [{ ...buildSearchResponseBody()[0] }];
    mockedFetchEligible.mockResolvedValueOnce(mockedResponse);

    const response = await POST(request);
    expect(response.status).toEqual(200);
    const body = await response.json();
    expect(body).toMatchSnapshot();
  });

  it('should return an error 500 when LCA fetching throws', async () => {
    const payload = new FormData();
    payload.append('beneficiaryLastname', 'DUPOND');
    payload.append('beneficiaryFirstname', 'MANON');
    payload.append('beneficiaryBirthDate', '2011-01-01');
    payload.append('recipientResidencePlace', '05024');

    const request = new Request('http://localhost/api/eligibility-test/search', {
      method: 'POST',
      body: payload,
    });

    const mockedFetchQrCode = fetchEligible as jest.Mock;
    mockedFetchQrCode.mockImplementationOnce(() => {
      throw new Error();
    });

    const response = await POST(request);
    expect(response.status).toEqual(500);
  });
});
