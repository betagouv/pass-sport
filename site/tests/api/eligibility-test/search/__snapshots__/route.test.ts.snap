// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /eligibility-test/search should return fetched data 1`] = `
[
  {
    "date_naissance": "2011-01-01T23:00:00.000Z",
    "id": 123,
    "matricule": "9999999999999",
    "nom": "DUPOND",
    "organisme": "MSA",
    "prenom": "MANON",
    "situation": "jeune",
  },
]
`;
